src/coordination/coordinator_instance.cpp

```diff
@@ -166,8 +102,42 @@ auto CoordinatorInstance::TryFailover() -> void {
     return;
   }
 
-  // TODO: Smarter choice
-  auto new_main = ranges::begin(alive_replicas);
+  std::string most_up_to_date_instance;
+  std::optional<uint64_t> latest_commit_timestamp;
+  std::optional<std::string> latest_epoch;
+  {
+    // for each DB in instance we get one DatabaseHistory
+    using DatabaseHistories = replication_coordination_glue::DatabaseHistories;
+    std::vector<std::pair<std::string, DatabaseHistories>> instance_database_histories;
+
+    bool success{true};
+    std::for_each(alive_replicas.begin(), alive_replicas.end(),
+                  [&success, &instance_database_histories](ReplicationInstance &replica) {
+                    if (!success) {
+                      return;
+                    }
+                    auto res = replica.GetClient().SendGetInstanceTimestampsRpc();
+                    if (res.HasError()) {
+                      spdlog::error("Could get per db history data for instance {}", replica.InstanceName());
+                      success = false;
+                      return;
+                    }
+                    instance_database_histories.emplace_back(replica.InstanceName(), std::move(res.GetValue()));
+                  });
+
+    if (!success) {
+      spdlog::error("Aborting failover as at least one instance didn't provide per database history.");
+      return;
+    }
+
+    most_up_to_date_instance =
+        coordination::ChooseMostUpToDateInstance(instance_database_histories, latest_epoch, latest_commit_timestamp);
```

Comment:
I don't think this should be utils method since it is used only once. It is nice to separate into function but function should then be here as private method not as util since it won't be used by any other module

Reply:
I see what you mean, I will try to push it as static private member function. I think it was util only because it was easier to unit test at first.

Reply:
Done